source/Halibut/Transport/Streams/NetworkTimeoutStream.cs

```diff
@@ -139,16 +139,49 @@ void ThrowMeaningfulException(Exception? innerException = null)
         public override void Close() => inner.Close();
 
         public override async Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken) => await inner.CopyToAsync(destination, bufferSize, cancellationToken);
+        
+        public override int ReadByte() => inner.ReadByte();
 
-        public override int EndRead(IAsyncResult asyncResult) => inner.EndRead(asyncResult);
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+        {
+            // BeginRead does not respect timeouts. So force it to use ReadAsync, which does.
+            return ReadAsync(buffer, offset, count, CancellationToken.None).AsAsynchronousProgrammingModel(callback, state);
+        }
 
-        public override void EndWrite(IAsyncResult asyncResult) => inner.EndWrite(asyncResult);
+        public override int EndRead(IAsyncResult asyncResult)
+        {
+            try
+            {
+                return ((Task<int>)asyncResult).Result;
+            }
+            catch (AggregateException e) when (e.InnerExceptions.Count == 1 && e.InnerException is not null)
```

Comment:
It always comes back as an `AggregateException` (even though it's just 1). So this helps ensure we throw the correct exception, rather than having the caller need to unwrap it.

Reply:
Was this caught by the tests?

Reply:
Yep!